File: most-adapter/src/main/java/com/softwareverde/mostadapter/MostFunctionStereotype.java
package com.softwareverde.mostadapter;

import java.util.ArrayList;
import java.util.List;

public class MostFunctionStereotype {
    private String _name;
    private boolean _supportsNotification;
    private String _category;
    private List<Operation> _operations;

    public String getName() {
        return _name;
    }

    public void setName(String name) {
        _name = name;
    }

    public boolean supportsNotification() {
        return _supportsNotification;
    }

    public void setSupportsNotification(boolean supportsNotification) {
        _supportsNotification = supportsNotification;
    }

    public String getCategory() {
        return _category;
    }

    public void setCategory(String category) {
        _category = category;
    }

    public List<Operation> getOperations() {
        return new ArrayList<>(_operations);
    }

    public void addOperation(final Operation operation) {
        _operations.add(operation);
    }

    public void setOperations(final List<Operation> operations) {
        _operations = new ArrayList<>(operations);
    }
}

File: app/src/main/webapp/js/classes/most-function-stereotype.js
class MostFunctionStereotype {
    static fromJson(json) {
        const mostFunctionStereotype = new MostFunctionStereotype();

        const operations = [];
        const operationsJson = json.operations;
        for (let i in operationsJson) {
            const operationJson = operationsJson[i];
            const operation = Operation.fromJson(operationJson);
            operations.push(operation);
        }

        mostFunctionStereotype.setId(json.id);
        mostFunctionStereotype.setName(json.name);
        mostFunctionStereotype.setSupportsNotification(json.supportsNotification);
        mostFunctionStereotype.setCategory(json.category);
        mostFunctionStereotype.setOperations(operations);

        return mostFunctionStereotype;
    }

    static toJson(mostFunctionStereotype) {
        const jsonMostFunctionStereotype = {
            id:                     mostFunctionStereotype.getId(),
            name:                   mostFunctionStereotype.getName(),
            supportsNotification:   mostFunctionStereotype.getSupportsNotification(),
            category:               mostFunctionStereotype.getCategory(),
        };

        // Jsonify operations array, which simply contains ids for operations.
        const operations = mostFunctionStereotype.getOperations();
        const operationsJson = [];
        for (let i in operations) {
            const operationJson = Operation.toJson(operations[i]);
            operationsJson.push(operationJson);
        }
        jsonMostFunctionStereotype.operations = operationsJson;

        return jsonMostFunctionStereotype;
    }

    constructor() {
        this._id                        = null;
        this._name                      = "";
        this._supportsNotification      = null;
        this._category                  = "";
        this._operations                = [];
    }

    setId(id) {
        this._id = id;
    }

    getId() {
        return this._id;
    }

    setName(name) {
        this._name = name;
    }

    getName() {
        return this._name;
    }

    setSupportsNotification(supportsNotification) {
        this._supportsNotification = supportsNotification;
    }

    getSupportsNotification() {
        return this._supportsNotification;
    }

    setCategory(category) {
        this._category = category;
    }

    getCategory() {
        return this._category;
    }

    setOperations(operations) {
        this._operations = operations;
    }

    getOperations() {
        return this._operations;
    }
}